styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java

```diff
@@ -91,72 +104,110 @@ void tick() {
       return;
     }
 
-    final Map<String, Long> backfillStates = getBackfillStates();
-    backfills.forEach(backfill -> guard(() -> triggerBackfill(backfill, backfillStates)).run());
+    shuffler.accept(backfills);
+
+    backfills.forEach(backfill -> guard(() -> triggerAndProgress(backfill)).run());
 
     final long durationMillis = t0.until(time.get(), ChronoUnit.MILLIS);
     stats.recordTickDuration(TICK_TYPE, durationMillis);
   }
 
-  private void triggerBackfill(Backfill backfill, Map<String, Long> backfillStates) {
+  private void triggerAndProgress(Backfill backfill) {
     final Optional<Workflow> workflowOpt = workflowCache.workflow(backfill.workflowId());
     if (!workflowOpt.isPresent()) {
-      LOG.warn("workflow not found for backfill, skipping rest of triggers: {}", backfill);
-      final BackfillBuilder builder = backfill.builder();
-      builder.halted(true);
-      storeBackfill(builder.build());
+      LOG.debug("workflow not found for backfill {}, halt it.", backfill);
+      storeBackfill(backfill.builder().halted(true).build());
       return;
     }
 
     final Workflow workflow = workflowOpt.get();
 
+    // this is best effort because querying active states is not strongly consistent so the
+    // initial remaining capacity may already be wrong
     final int remainingCapacity =
-        backfill.concurrency() - backfillStates.getOrDefault(backfill.id(), 0L).intValue();
+        backfill.concurrency() - stateManager.activeStatesByTriggerId(backfill.id()).size();
 
-    Instant partition = backfill.nextTrigger();
+    if (remainingCapacity < 1) {
+      LOG.debug("No capacity left for backfill {}", backfill);
+      return;
+    }
 
-    for (int i = 0; i < remainingCapacity && partition.isBefore(backfill.end()); i++) {
+    final Instant initialNextTrigger = backfill.nextTrigger();
+    while (true) {
       try {
-        final CompletableFuture<Void> processed = triggerListener
-            .event(workflow, Trigger.backfill(backfill.id()), partition)
-            .toCompletableFuture();
-        // Wait for the trigger execution to complete before proceeding to the next partition
-        processed.get();
-      } catch (AlreadyInitializedException ignored) {
-        // nop
-      } catch (Throwable e) {
-        LOG.warn("failed to trigger backfill for state [{}]: {}", partition, backfill, e);
-        return;
+        if (!storage.runInTransaction(tx ->
+            triggerNextPartitionAndProgress(tx, backfill.id(), workflow,
+                initialNextTrigger, remainingCapacity))) {
+          break;
+        }
+      } catch (IOException e) {
+        // transaction failure, yield
+        LOG.debug("Transaction failure when working on backfill {}", backfill, e);
```

Comment:
IOException can be something else than transaction failure; isn't this a potentially misleading message?